### DISCUSSION TITLE
Live plotting/ Draw objects in scene

### DISCUSSION THREAD

**Discussion by v-r-a** [google-deepmind/mujoco#446] [Asking for Help] [answered]
I am not able to trace the source code corresponding to the following functionalities in `simulate.cc`
![simulate_menu_cannot_find_code](https://user-images.githubusercontent.com/93522619/185400817-d5a15823-9a14-4355-8c45-d376b626c1c0.png)

I am interested in live plotting-- 
points like the centre of mass (already available in `simulate`) , the centre of pressure,
desired trajectory,
inertia forces etc 
real time in the simulation.

**Comment by yuvaltassa** [comment]
IIUC you mean that you want to draw 3D objects directly in the scene, correct?
This is possible and easy. Are you using C++ or Python?

**Reply by v-r-a** [reply]
Yes. 
I'm using C++

**Comment by yuvaltassa** [comment, marked as answer]
In that case you can follow the patterns in [`engine_vis_visualize.c`](https://github.com/deepmind/mujoco/blob/main/src/engine/engine_vis_visualize.c). 

The place to follow down from `simulate`, btw, is the [`mjv_updateScene`](https://github.com/deepmind/mujoco/blob/main/simulate/simulate.cc#L1614) call.

There is a fairly straighforward `mjvScene` struct, you add geoms to it and they get rendered. Please excuse the abnormally long `mjv_addGeoms` function... we should break it up 🙂

**Reply by v-r-a** [reply]
Okay, I'll try this. 
Thank you

**Reply by v-r-a** [reply]
@yuvaltassa, I could render my own centre of mass visual by inserting the following code (copied and adapted from `START` and `FINISH` macros used in `mjv_addGeom`) in `basic.cc` (also in `simulate.cc`). I hope this is the intended way.

```
    // add a decorative geometry
    mjvGeom* mygeom;
    mjtNum sphrad[3] = {0.05,0,0};
    mjtNum sphpos[3] = {0.,0.,0.};
    mjtNum myid3x3[9] = {1,0,0,0,1,0,0,0,1};
    float sphrgba[4] = {0.,0.,0.9,0.8};
    sphpos[0] = d->subtree_com[0];
    sphpos[1] = d->subtree_com[1];
    sphpos[2] = d->subtree_com[2];

    // update scene and render
    mjv_updateScene(m, d, &opt, NULL, &cam, mjCAT_ALL, &scn);
    printf("Number of geometries in the buffer AA: %d\n",scn.ngeom);
    //one more geom to render
    scn.ngeom = scn.ngeom +1;
    // mygeom now points to the last location in geoms buffer
    mygeom = scn.geoms + scn.ngeom -1;
    // decor geom
    mygeom->objtype = mjOBJ_UNKNOWN;
    mygeom->objid = -1;
    mygeom->category = mjCAT_DECOR;
    mygeom->segid = scn.ngeom;  
    mjv_initGeom(mygeom,mjGEOM_SPHERE,sphrad,sphpos,myid3x3,sphrgba);
    printf("Number of geometries in the buffer BB: %d\n",scn.ngeom);
    
    mjr_render(viewport, &scn, &con);
```
